File: Data Structures & Algorithms/target-sum/submission-6.cpp
class Solution {
public:
    map<pair<int,int>,int> d;
    int dfs(int i,int cur,vector<int>& nums, int target){
        pair<int,int> key = {i, cur};
        if(d.count(key)) return d[key] ;
            
        else if(i == nums.size()-1){
            if(cur==target)
              return 1;
            else return 0;
        }
        d[key]=dfs(i+1,nums[i+1]+cur,nums,target)+dfs(i+1,cur-nums[i+1],nums,target);
        return d[key] ;
    }
    int findTargetSumWays(vector<int>& nums, int target) {
        return dfs(0,nums[0],nums,target)+dfs(0,-nums[0],nums,target);
    }
};

File: Data Structures & Algorithms/regular-expression-matching/submission-6.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        n,m=len(s),len(p)
        def dfs(i,j):
            if j==m:
                return i==n
            if i<n and ( s[i]==p[j] or p[j]=='.' ):
                match=True
            else:
                match = False
            if j+1< m and p[j+1]=="*" :
                return  dfs(i,j+2) or  ( match and dfs(i+1,j) ) 
            if match :
                return dfs(i+1,j+1)
            return False
        return dfs(0,0)
                
            
            
            